File: scala_word_count/src/main/java/com/alex/test/flink/WordCount.java
package com.alex.test.flink;

import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.api.java.utils.ParameterTool;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.util.Collector;
import scala.Tuple2;

/**
 * @ClassName WordCount
 * @Description TODO
 * @Author Alex
 * @Date 2019/2/19 17:30
 * @Version 1.0
 **/
public class WordCount {

    public static void main(String[] args) {

        final ParameterTool params = ParameterTool.fromArgs(args);

        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

        env.getConfig().setGlobalJobParameters(params);

        DataStream<String> text;
        if (params.has("input")) {
            text = env.readTextFile(params.get("input"));
        } else {
            text = env.fromElements(WordCountData.WORDS);
        }

        DataStream<Tuple2<String, Integer>> counts = text.flatMap(new Tokenizer())
                .keyBy(0).sum(1);

        if (params.has("input")) {
            counts.writeAsText(params.get("output"));
        } else {
            counts.print();
        }

    }

    public static final class Tokenizer implements FlatMapFunction<String, Tuple2<String, Integer>> {
        @Override
        public void flatMap(String value, Collector<Tuple2<String, Integer>> collector) throws Exception {
            String[] tokens = value.toLowerCase().split("\\W+");

            for (String token : tokens) {
                if (token.length() > 0) {
                    collector.collect(new Tuple2<>(token, 1));
                }
            }
        }
    }

}

File: scala_word_count/src/main/java/com/alex/test/WordCount.scala
package com.alex.test

import org.apache.flink.api.scala._
import org.apache.flink.core.fs.FileSystem.WriteMode

object WordCount {
  def main(args: Array[String]): Unit = {
    val env = ExecutionEnvironment.createLocalEnvironment(1)

    // 从本地读取文件
    val text = env.readTextFile("C:/document/cdr/20180505-1556.log")

    // 单词统计
//    val counts = text.flatMap { _.toLowerCase.split("\\W+") filter { _.nonEmpty } }
//      .map { (_, 1) }
//      .groupBy(0)
//      .sum(1)

    val counts = text.flatMap(_.toLowerCase.split("\\W+"))
        .map((_, 1)).groupBy(0).sum(1)

    //输出结果
    counts.print()

    //保存文件到txt
    counts.writeAsText("c:/output.txt", WriteMode.OVERWRITE)
    env.execute("Scala WordCount Example")
  }

}